students/andykwon/lesson05/mailroom_pt3.py

```diff
@@ -0,0 +1,188 @@
+                print(name)
+
+            continue
+
+        elif response == "3":
+            return
+
+        # Can the above if/else be replaced with setdefault?? look into it...
+
+
+def donations(keys):
+    return keys[1]
+
+
+def report():
+    """
+    prints a report of the donors and their contributions
+    """
+    print("== Generate a Report ==\n")
+
+    report = []
+    report_data = []
+
+    header = ("{:<25s} | {:^15s} | {:^10s} | {:15s}".format(
+        "Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+    divider = "_" * len(header)
+
+    report.append(header)
+    report.append(divider)
+
+    for key_name in donors:
+        name = key_name
+        total_given = sum(donors[key_name])
+        total_count = len(donors[key_name])
+        avg_given = total_given / total_count
+        report_data.append((name, total_given, total_count, avg_given))
+
+    # Sort with respect to "Total Given" value.
+    report_data.sort(key=donations, reverse=True)
+
+    # Adding the data into report with formatting
+    for data in report_data:
+        report.append("{:25s}   ${:14.2f}   {:10d}   ${:11.2f}".format(* data))
+
+    print("\n".join(report))
+    return "\n".join(report)
+
+
+def letters():
+
+    print("== Send letters to everyone ==")
+
+    for key_name in donors:
+        file = open("Letter_for_" + key_name + ".txt", "w")
+
+        letter = ("Dear " + key_name + ",\n\n" +
+                  "Thank you for your recent donation of $" +
+                  "%.2f" % donors[key_name][-1] + ".\n\n" +
+                  "It will be put to very good use.\n\n" +
+                  "Sincerely,\n\n" + "-The Team")
+
+        file.write(letter)
+
+        file.close()
+
+
+def quit():
+
+    while True:
+
+        print("Are you sure you want to quit?\n"
+              "1) Yes, quit\n"
+              "2) Back\n")
+
+        response = input(">")
+
+        # Test to see if input is an option
+        try:
+            if int(response) > 2:
+                raise ValueError
+        except ValueError:
+            print("invalid input\n")
+
+        if response == "1":
+            sys.exit()
+        elif response == "2":
+            return
+
+
+def mainloop():
+
+    response = ""
+
+    while True:
+        print("Welcome to Mailroom")
+        print("what do you want to do?")
+        print("1) thank you\n"
+              "2) report\n"
+              "3) Send letters to everyone\n"
+              "4) quit\n")
+
+        response = input(">>")
+
+        # Test to see if input is an option
+        try:
+            if int(response) > 4:
+                raise ValueError
+        except ValueError:
+            print("invalid input\n")
+
+        if response == "1":
+            thank_you()
+            continue
+        elif response == "2":
+            report()
+            continue
+        elif response == "3":
+            letters()
+            continue
+        elif response == "4":
```

Comment:
Instead of writing a ton of elifs it'd be best to use a dictionary for switch statements like this. It's more efficient for the computer, and less typing.